[11.0][WIP] stock_barcodes: Allow to read barcodes from pickings

sergio-teruel opened the pull request:
cc @Tecnativa

sergio-teruel committed:
[ADD] stock_barcodes: Allow to read barcodes from pickings

pedrobaeza commented on stock_barcodes/README.rst:
```diff
@@ -49,6 +52,24 @@ To use the barcode interface on inventory:
 
+Barcode interface for picking operations
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+You can use the barcode interface in a picking or an operation type, the mainly
```
s/mainly/main

pedrobaeza commented on stock_barcodes/README.rst:
```diff
@@ -49,6 +52,24 @@ To use the barcode interface on inventory:
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+You can use the barcode interface in a picking or an operation type, the mainly
+difference is that if you open the barcode interface from a picking, this
+picking is locked and you read product for it.
```
s/product/products

pedrobaeza commented on stock_barcodes/README.rst:
```diff
@@ -49,6 +52,24 @@ To use the barcode interface on inventory:
+
+#. Go to *Inventory*.
+#. Click on scanner button on any operation type.
+#. Start read barcodes.
```
s/read/reading

pedrobaeza commented on stock_barcodes/README.rst:
```diff
@@ -49,6 +52,24 @@ To use the barcode interface on inventory:
+
+#. Go to *Inventory > Transfers*.
+#. Click to "Scan barcodes" smart button.
+#. Start read barcodes.
```
s/read/reading

pedrobaeza commented on stock_barcodes/__manifest__.py:
```diff
@@ -3,7 +3,7 @@
     "name": "Stock Barcodes",
     "summary": "It provides read barcode on stock operations.",
-    "version": "11.0.1.0.0",
+    "version": "11.0.1.1.0",
```
This huge change deserves `11.0.2.0.0`

pedrobaeza commented on stock_barcodes/README.rst:
```diff
@@ -103,9 +124,18 @@ User can remove the last read scan.
-* Barcode interface for reading barcodes on pickings.
 * Excute action_done() method outside onchange environment.
 * Allow create product when a barcode has not been found.
+* Select picking with barcode.
```
Allow to select picking reading its barcode.

pedrobaeza commented on stock_barcodes/README.rst:
```diff
@@ -103,9 +124,18 @@ User can remove the last read scan.
 ======================
 
-* Barcode interface for reading barcodes on pickings.
 * Excute action_done() method outside onchange environment.
```
Add a line about selecting multiple pickings at the same time.

pedrobaeza commented on stock_barcodes/models/stock_picking.py:
```diff
@@ -0,0 +1,42 @@
+        return action
+
+
+class PickingType(models.Model):
```
Put it in a separate file and with convention class name

pedrobaeza commented on stock_barcodes/static/description/index.html:
```diff
@@ -3,7 +3,7 @@
 <head>
 <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
-<meta name="generator" content="Docutils 0.15.1: http://docutils.sourceforge.net/" />
+<meta name="generator" content="Docutils 0.14: http://docutils.sourceforge.net/" />
```
Update your local docutils to 0.15.1 if you are pre-generating the README/index

pedrobaeza commented on stock_barcodes/wizard/stock_barcodes_read_picking.py:
```diff
@@ -0,0 +1,293 @@
+        return [
+            (rec.id, '{} - {} - {}'.format(
+                _('Barcode reader'),
+                rec.picking_id.name, self.env.user.name)) for rec in self]
```
If not `picking_id` (starting from picking type), it should be reflected in the initial name.

pedrobaeza commented on stock_barcodes/wizard/stock_barcodes_read_picking.py:
```diff
@@ -0,0 +1,293 @@
+                rec.picking_id.name, self.env.user.name)) for rec in self]
+
+    def _set_default_picking(self):
+        picking_id = self.env.context.get('default_picking_id', False)
```
Remove this method and see below on `create` method.

pedrobaeza commented on stock_barcodes/wizard/stock_barcodes_read_picking.py:
```diff
@@ -0,0 +1,293 @@
+        # When user click any view button the wizard record is create and the
+        # picking candidates have been lost, so we need set it.
+        wiz = super().create(vals)
+        wiz._set_default_picking()
```
```suggestion
        if wiz.picking_id:
            wiz._set_candidate_pickings(wiz.picking_id)
```

pedrobaeza commented on stock_barcodes/wizard/stock_barcodes_read_picking.py:
```diff
@@ -0,0 +1,293 @@
+        # Add to candidate pickings the default picking. We are in a wizard
+        # view, so for create a candidate picking with the same default picking
+        # we need create it in this onchange
+        self._set_default_picking()
```
An onchange **musn't** create records. You should switch to a compute virtual many2many field.

pedrobaeza commented on stock_barcodes/wizard/stock_barcodes_read_picking.py:
```diff
@@ -0,0 +1,293 @@
+        self._set_default_picking()
+
+    def action_done(self):
+        if self.check_done_conditions():
```
why not a single if with an `and`?

pedrobaeza commented on stock_barcodes/wizard/stock_barcodes_read_picking.py:
```diff
@@ -0,0 +1,293 @@
+            candidate_pickings = moves_todo.mapped('picking_id')
+            candidate_pickings_count = len(candidate_pickings)
+            if candidate_pickings_count > 1:
+                self.select_cadidate_picking(candidate_pickings)
```
```suggestion
                self.select_candidate_picking(candidate_pickings)
```
why this sub-methods? Are they really needed?

pedrobaeza commented on stock_barcodes/wizard/stock_barcodes_read_picking.py:
```diff
@@ -0,0 +1,293 @@
+        if not self._search_candidate_pickings(moves_todo):
+            return False
+        lines = moves_todo.mapped('move_line_ids').filtered(
+            lambda l: (l.picking_id == self.picking_id and
```
Why not adding this to the initial domain prepared by `_prepare_stock_moves_domain`?

sergio-teruel replied:
Mybe I only need filter by lot

pedrobaeza commented on stock_barcodes/wizard/stock_barcodes_read_picking.py:
```diff
@@ -0,0 +1,293 @@
+                       l.product_id == self.product_id and
+                       l.lot_id == self.lot_id))
+        move_line = StockMoveLine.browse()
+        if lines:
```
Better algorithm IMO + cover the case where the qty is higher than the existing line creating an extra sml instead of putting it in the last one.

```suggestion
        available_qty = self.product_qty
        for line in lines:
            assigned_qty = min(line.product_uom_qty - line.qty_done, available_qty)
            line.qty_done += assigned_qty
            available_qty -= assigned_qty
            if float_compare(available_qty, 0, ...) < 0:
                break
        if float_compare(available_qty, 0, ...) > 0:
            # Create an extra stock move line if this product has an
            ...
```

pedrobaeza commented on stock_barcodes/wizard/stock_barcodes_read_picking.py:
```diff
@@ -0,0 +1,293 @@
+        self.picking_product_qty = sum(
+            move_lines.mapped('move_id.quantity_done'))
+
+    def select_cadidate_picking(self, picking_candidates):
```
Why this "ghost" method?

pedrobaeza commented on stock_barcodes/wizard/stock_barcodes_read_picking.py:
```diff
@@ -0,0 +1,293 @@
+            ('product_id', '=', log_scan.product_id.id),
+            ('lot_id', '=', log_scan.lot_id.id),
+            ('move_id.picking_id.picking_type_id.code', '=',
+             self.env.context.get('default_picking_type_code')),
```
You should take `self.picking_type_code`, isn't it?

sergio-teruel replied:
Yes..

pedrobaeza commented on stock_barcodes/wizard/stock_barcodes_read_picking.py:
```diff
@@ -0,0 +1,293 @@
+        log_scan = first(self.scan_log_ids.filtered(
+            lambda x: x.create_uid == self.env.user))
+        if log_scan:
+            domain = self._prepare_move_line_undo_domain(log_scan=log_scan)
```
Not sure if this is going to return always your last scan or others. The only way to be sure is to add a scan_log_id at sml level.

pedrobaeza commented on stock_barcodes/wizard/stock_barcodes_read_picking.py:
```diff
@@ -0,0 +1,293 @@
+            for line in lines:
+                qty = line.qty_done - log_scan.product_qty
+                undo_qty += line.qty_done
+                line.qty_done = qty if qty > 0.0 else 0.0
```
```suggestion
                line.qty_done = max(qty, 0.0)
```

pedrobaeza commented on stock_barcodes/wizard/stock_barcodes_read_picking.py:
```diff
@@ -0,0 +1,293 @@
+                qty = line.qty_done - log_scan.product_qty
+                undo_qty += line.qty_done
+                line.qty_done = qty if qty > 0.0 else 0.0
+                if undo_qty == log_scan.product_qty:
```
`float_compare`

pedrobaeza commented on stock_barcodes/wizard/stock_barcodes_read_picking.py:
```diff
@@ -0,0 +1,293 @@
+
+    def _get_wizard_barcode_read(self):
+        return self.env['wiz.stock.barcodes.read.picking'].browse(
+            self.env.context['wiz_read_scan_id'])
```
Use `wiz_barcode_id` field instead

sergio-teruel committed:
[ADD] stock_barcodes: Allow to read barcodes from pickings. Last changes

PhatDang commented on stock_barcodes/models/stock_barcodes_read_log.py:
```diff
@@ -43,3 +43,46 @@ class StockBarcodesReadLog(models.Model):
     )
+    picking_id = fields.Many2one(
+        comodel_name='stock.picking',
+        string='Picking',
```
Odoo will call the name of the field instead of a string of name. So you don't need to define it.
In this case, if `the name of field = the name of string`

pedrobaeza replied:
Well, although true, it's not required to save that argument. Any way, Python zen is "better explicit than implicit", so I think this can be kept.

PhatDang commented on stock_barcodes/wizard/stock_barcodes_read_picking.py:
```diff
@@ -0,0 +1,343 @@
+
+    picking_id = fields.Many2one(
+        comodel_name='stock.picking',
+        string='Picking',
```
If the same name between with the name of the field and the string name. 
Odoo will call it from the name of the field. You do not need to define the string name.

PhatDang commented on stock_barcodes/wizard/stock_barcodes_read_picking.py:
```diff
@@ -0,0 +1,343 @@
+    candidate_picking_ids = fields.One2many(
+        comodel_name='wiz.candidate.picking',
+        inverse_name='wiz_barcode_id',
+        string='Candidate pickings',
```
Same above

pedrobaeza replied:
In this case, the label would be `Candidate Picking`, so it's not the same as the wanted one.

PhatDang commented on stock_barcodes/wizard/stock_barcodes_read_picking.py:
```diff
@@ -0,0 +1,343 @@
+    )
+    picking_id = fields.Many2one(
+        comodel_name='stock.picking',
+        string='Picking',
```
Odoo will call it from the name of the field. You do not need to define the string name.

PhatDang commented on stock_barcodes/wizard/stock_barcodes_read_picking.py:
```diff
@@ -0,0 +1,343 @@
+        comodel_name='res.partner',
+        related='picking_id.partner_id',
+        readonly=True,
+        string='Partner',
```
Same above

PhatDang commented on stock_barcodes/wizard/stock_barcodes_read_picking.py:
```diff
@@ -0,0 +1,343 @@
+        readonly=True,
+    )
+    product_uom_qty = fields.Float(
+        'Demand', compute='_compute_picking_quantity',
```
Please break line for `compute=`
https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#idioms

pedrobaeza replied:
Well, guidelines doesn't specify this strictly, but it's harmless as well. We plan to introduce auto-formatting soon, so this kind of discussions will be outdated :) If you are coming to OCA days, you can join the track.

PhatDang commented on stock_barcodes/wizard/stock_barcodes_read_picking.py:
```diff
@@ -0,0 +1,343 @@
+        readonly=True,
+    )
+    product_qty_done = fields.Float(
+        'Done', compute='_compute_picking_quantity',
```
Same above

PhatDang reviewed (commented):
Please take a look at my comment. Thank you!

pedrobaeza commented:
@PhatDang thanks any way for your review.

@sergio-teruel Travis is red

rafaelbn reviewed (approved):
:+1:  all bots :green_apple: functional working

AdriaGForgeFlow reviewed (commented):
Doing an inventory adjustment and after scanning some products you exit the wizard and click the button that sets the quantities of all lines to 0. After that you go into the wizard again and scan one product that was scanned before setting quantities to 0. All previous logs appear and, even if it does not crash, you can click the button to remove last scanned until the logs are empty. 
Is this the desired behavior?
![stock_barcodes](https://user-images.githubusercontent.com/44768500/66493183-e3013c80-eab5-11e9-8596-c799e98955ce.gif)

Just to know since I'm doing a backport to v10.

sergio-teruel commented:
Hi @ageficent ,

> Is this the desired behavior?
Yes, if user really has been scanned the barcode and if other user or himself go to inventory and reset units the scanning logs is the same. (Is my opinion).
In the lastest PR's commit I add a deatil scanning log. mybe in this lines we can add scanned barcoed and units but in any case we will remove log lines when a inventory line has been removed.. (for performance)

AdriaGForgeFlow commented:
> Hi @ageficent ,
> 
> > Is this the desired behavior?
> > Yes, if user really has been scanned the barcode and if other user or himself go to inventory and reset units the scanning logs is the same. (Is my opinion).
> > In the lastest PR's commit I add a deatil scanning log. mybe in this lines we can add scanned barcoed and units but in any case we will remove log lines when a inventory line has been removed.. (for performance)

Thanks for the answer!

Another issue that I found, I think it's the same that appeared in the OCA days. In a picking whenever you want to validate the picking from the wizard once you have scanned all products, the first time the following error appears:
![image](https://user-images.githubusercontent.com/44768500/66564926-79d70300-eb61-11e9-9f39-a9d996a1bbc3.png)

If you dismiss the error and try to validate again it will work.
![stock_barcodes_picking](https://user-images.githubusercontent.com/44768500/66564947-85c2c500-eb61-11e9-8b8f-32fa7fcf120b.gif)

sergio-teruel commented:
Yes yes, i know...

sergio-teruel committed:
[FIX] stock_barcodes: Fix validate picking from candidate picking kanban view

sergio-teruel commented:
@ageficent Fix done!! Thanks for your help :smile:

AdriaGForgeFlow commented:
> @ageficent Fix done!! Thanks for your help

Perfect! Now it works like a charm :+1: 

Are all the functionalities that you wanted to add to the module already done? Or do you plan on adding more features?

pedrobaeza commented:
We don't plan to add more for now. Do you have something in mind?

AdriaGForgeFlow commented:
> We don't plan to add more for now. Do you have something in mind?

No, I believe this is already a very nice improvement. Thanks for the work!

AdriaGForgeFlow reviewed (approved):
Functional test and code review :+1:

pedrobaeza commented:
@chienandalu give your final blessing after customer's tests and good to go

chienandalu reviewed (commented):
Maybe we shouldn't allow to enter lots when the product has no tracking:

![image](https://user-images.githubusercontent.com/5040182/66662651-82a30400-ec49-11e9-9354-6b6e5ab655b3.png)

Not a big deal (the quants are set without such lot), although maybe a little bit misleading to some users.

sergio-teruel commented:
But these behavior os not present un Odio... Aldo, other feature is that user can scan a lot barcode directly and if this lot only has one product you do not need scan the product... 
The lot fieles only is hidden if user has not stock_production_lot security group

pedrobaeza commented:
@sergio-teruel in standard, with non detailed operations, you don't see the button for opening the lot popup, so it's kind of similar of this. If you scan directly a lot, then in that moment both product and lot will be filled, and the lot will be shown. I vote also for showing it if no product or product has tracking (and of course with the security group).

sergio-teruel committed:
[IMP] stock_barcodes: Hidden lot for non tracking products.
When user creates a lot, set default product the wizard selected product.
When confirm the lot creation wizard, set this lot in lot wizard scanning barcode field.

sergio-teruel commented:
Ok, OK... Changes done!
Other improvements in this commit:
When user creates a lot, set default product the wizard selected product.
When confirm the lot creation wizard, set this lot in lot wizard scanning barcode field.

pedrobaeza commented:
OK, thanks for the final changes. Let's merge this! /ocabot merge

![66479506-efc56680-ea9c-11e9-9103-82f5919e7293](https://user-images.githubusercontent.com/7165771/66703352-01637400-ed12-11e9-913f-8320ecf94e78.gif)
